File: backend/app/routers/timeline.py
"""Timeline router — unified chronological feed of all entity types."""

from __future__ import annotations

from datetime import date

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.database import get_db
from app.schemas.timeline import TimelineResponse
from app.services.timeline import get_timeline

router = APIRouter(prefix="/api", tags=["timeline"])


@router.get("/timeline", response_model=TimelineResponse)
def timeline(
    page: int = Query(1, ge=1, description="Page number"),
    per_page: int = Query(20, ge=1, le=100, description="Items per page"),
    types: str | None = Query(
        None, description="Comma-separated entity types: journal,metric,result,goal"
    ),
    tag: str | None = Query(None, description="Filter by tag"),
    date_from: date | None = Query(None, description="Start date (inclusive)"),
    date_to: date | None = Query(None, description="End date (inclusive)"),
    db: Session = Depends(get_db),  # noqa: B008
) -> TimelineResponse:
    """Return a unified, chronological timeline of health data."""
    entity_types = [t.strip() for t in types.split(",") if t.strip()] if types else None

    result = get_timeline(
        db,
        page=page,
        per_page=per_page,
        types=entity_types,
        tag=tag,
        date_from=date_from,
        date_to=date_to,
    )

    return TimelineResponse(**result)

File: frontend/src/api/timeline.ts
import { api } from "./client";
import type { TimelineResponse } from "../types/timeline";

export function getTimeline(params?: {
  page?: number;
  per_page?: number;
  types?: string;
  tag?: string;
  date_from?: string;
  date_to?: string;
}): Promise<TimelineResponse> {
  const search = new URLSearchParams();
  if (params?.page) search.set("page", String(params.page));
  if (params?.per_page) search.set("per_page", String(params.per_page));
  if (params?.types) search.set("types", params.types);
  if (params?.tag) search.set("tag", params.tag);
  if (params?.date_from) search.set("date_from", params.date_from);
  if (params?.date_to) search.set("date_to", params.date_to);
  const qs = search.toString();
  return api.get<TimelineResponse>(`/api/timeline${qs ? `?${qs}` : ""}`);
}
